src/akkudoktoreos/server/fastapi_server.py

```diff
@@ -72,41 +89,219 @@ class PdfResponse(FileResponse):
+    return PydanticDateTimeSeries.from_series(pdseries)
+
+
+@app.put("/v1/measurement/value")
+def fastapi_measurement_value_put(
+    datetime: Any, key: str, value: Union[float | str]
+) -> PydanticDateTimeSeries:
+    """Merge the measurement of given key and value into EOS measurements at given datetime."""
+    if key not in measurement_eos.record_keys:
+        raise HTTPException(status_code=404, detail=f"Key '{key}' not available.")
+    measurement_eos.update_value(datetime, key, value)
+    pdseries = measurement_eos.key_to_series(key=key)
+    return PydanticDateTimeSeries.from_series(pdseries)
+
+
+@app.put("/v1/measurement/series")
+def fastapi_measurement_series_put(
+    key: str, series: PydanticDateTimeSeries
+) -> PydanticDateTimeSeries:
+    """Merge measurement given as series into given key."""
+    if key not in measurement_eos.record_keys:
+        raise HTTPException(status_code=404, detail=f"Key '{key}' not available.")
+    pdseries = series.to_series()  # make pandas series from PydanticDateTimeSeries
+    measurement_eos.key_from_series(key=key, series=pdseries)
+    pdseries = measurement_eos.key_to_series(key=key)
+    return PydanticDateTimeSeries.from_series(pdseries)
+
+
+@app.put("/v1/measurement/dataframe")
+def fastapi_measurement_dataframe_put(data: PydanticDateTimeDataFrame) -> None:
+    """Merge the measurement data given as dataframe into EOS measurements."""
+    dataframe = data.to_dataframe()
+    measurement_eos.import_from_dataframe(dataframe)
+
+
+@app.put("/v1/measurement/data")
+def fastapi_measurement_data_put(data: PydanticDateTimeData) -> None:
+    """Merge the measurement data given as datetime data into EOS measurements."""
+    datetimedata = data.to_dict()
+    measurement_eos.import_from_dict(datetimedata)
+
+
+@app.get("/v1/prediction/keys")
+def fastapi_prediction_keys_get() -> list[str]:
+    """Get a list of available prediction keys."""
+    return sorted(prediction_eos.record_keys)
+
+
+@app.get("/v1/prediction/series")
+def fastapi_prediction_series_get(
+    key: str,
+    start_datetime: Optional[str],
+    end_datetime: Optional[str],
+) -> PydanticDateTimeSeries:
+    """Get prediction for given key within given date range as series.
+
+    Args:
+        start_datetime: Starting datetime (inclusive).
+            Defaults to start datetime of latest prediction.
+        end_datetime: Ending datetime (exclusive).
+            Defaults to end datetime of latest prediction.
+    """
+    if key not in prediction_eos.record_keys:
+        raise HTTPException(status_code=404, detail=f"Key '{key}' not available.")
+    if start_datetime is None:
+        start_datetime = prediction_eos.start_datetime
+    else:
+        start_datetime = to_datetime(start_datetime)
+    if end_datetime is None:
+        end_datetime = prediction_eos.end_datetime
+    else:
+        end_datetime = to_datetime(end_datetime)
+    pdseries = prediction_eos.key_to_series(key=key)
```

Comment:
pass datetimes

Reply:
This way a lot of datetime formats are possible. Using datetime would restrict that. Don't know what is better.

Reply:
sorry, i wasn't clear. key_to_series doesn't use the start/end_datetime from the rest params.

Reply:
The datetime parameters are converted by `to_datetime()`, which can convert a lot of different formats. `key_to_series()` does not use them, but they are used later on.